src/librustc/error_codes.rs

```diff
@@ -2126,6 +2126,74 @@ static X: u32 = 42;
 ```
 "##,
 
+E0728: r##"
+`await` has been used outside `async` function or block.
+
+Erroneous code examples:
+
+```edition2018,compile_fail,E0728
+# use std::pin::Pin;
+# use std::future::Future;
+# use std::task::{Context, Poll};
+
+# struct WakeOnceThenComplete(bool);
+
+# fn wake_and_yield_once() -> WakeOnceThenComplete {
+#     WakeOnceThenComplete(false)
+# }
+
+# impl Future for WakeOnceThenComplete {
+#     type Output = ();
+#     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+#         if self.0 {
+#             Poll::Ready(())
+#         } else {
+#             cx.waker().wake_by_ref();
+#             self.0 = true;
+#             Poll::Pending
+#         }
+#     }
+# }
+
+fn foo() {
+    wake_and_yield_once().await // `await` is used outside `async` context
+}
+```
+
+`await` is used to suspend the current computation until the given
+future is ready to produce a value. So it is legal only within
+an async context, like an `async fn` or an `async` block.
+
+```edition2018
+# use std::pin::Pin;
+# use std::future::Future;
+# use std::task::{Context, Poll};
+
+# struct WakeOnceThenComplete(bool);
+
+# fn wake_and_yield_once() -> WakeOnceThenComplete {
+#     WakeOnceThenComplete(false)
+# }
+
+# impl Future for WakeOnceThenComplete {
+#     type Output = ();
+#     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+#         if self.0 {
+#             Poll::Ready(())
+#         } else {
+#             cx.waker().wake_by_ref();
+#             self.0 = true;
+#             Poll::Pending
+#         }
+#     }
+# }
+
+async fn foo() {
+    wake_and_yield_once().await // `await` is used within `async` context
+}
+```
```

Comment:
Add an example inside of an `async` block, I'd say.

Reply:
Added an `async` block example, what about?

Reply:
Great.

My last nitpick would be to have a link to the async/await documentation, but this is already a great improvement.

Reply:
Hm, at glance, we don't have the links to keywords in error code's explanations. I found the links to RFCs or specific topics. We should have a link to the async/await?

Reply:
I would lean towards over linking over under linking, but I'll defer to @GuillaumeGomez

Reply:
I'd prefer to have the links indeed.

Reply:
Okay, the link should be https://doc.rust-lang.org/std/keyword.async.html and https://doc.rust-lang.org/std/keyword.await.html ?